routing.go

```diff
@@ -437,52 +371,39 @@ func (dht *IpfsDHT) Provide(ctx context.Context, key cid.Cid, brdcst bool) (err
 		// context is still fine, provide the value to the closest peers
 		// we managed to find, even if they're not the _actual_ closest peers.
 		if ctx.Err() != nil {
-			return ctx.Err()
+			return nil, ctx.Err()
 		}
 		exceededDeadline = true
 	case nil:
 	default:
-		return err
-	}
-
-	mes, err := dht.makeProvRecord(keyMH)
-	if err != nil {
-		return err
+		return nil, err
 	}
 
+	closestPeers := make([]peer.ID, 0, dht.bucketSize)
 	wg := sync.WaitGroup{}
 	for p := range peers {
+		closestPeers = append(closestPeers, p)
 		wg.Add(1)
 		go func(p peer.ID) {
 			defer wg.Done()
 			logger.Debugf("putProvider(%s, %s)", keyMH, p)
-			err := dht.sendMessage(ctx, p, mes)
+			err := dht.protoMessenger.PutProvider(ctx, p, keyMH, dht.host)
```

Comment:
This is slightly less efficient since we create the record `len(peers)` times. If that's an issue we can make a function that allows that, but this seems likely to be fine. Thoughts?